common/k8s/src/metrics_stats_aggregator.rs

```diff
@@ -0,0 +1,672 @@
+use anyhow::Result;
+use core::result::Result::Ok;
+
+use futures::{stream, Stream, StreamExt};
+use http::types::body::LineBuilder;
+
+use k8s_openapi::api::core::v1::{Container, ContainerStatus, Node, Pod};
+use kube::{
+    api::{Api, DynamicObject, GroupVersionKind, ListParams, ObjectList},
+    discovery, Client,
+};
+use serde_json::Value;
+
+use std::collections::HashMap;
+use std::time::Duration;
+use tokio::time::sleep;
+
+use crate::kube_stats::{
+    cluster_stats::ClusterStats,
+    container_stats::ContainerStats,
+    controller_stats::ControllerStats,
+    extended_pod_stats::ExtendedPodStats,
+    node_stats::{NodeContainerStats, NodePodStats, NodeStats},
+    pod_stats::PodStats,
+};
+
+pub struct MetricsStatsAggregator {
+    pub client: Client,
+}
+
+impl MetricsStatsAggregator {
+    pub fn new(client: Client) -> Self {
+        Self { client }
+    }
+
+    pub fn start_metrics_call_task(self) -> impl Stream<Item = LineBuilder> {
+        stream::unfold(self.client, |client| async {
+            sleep(Duration::from_millis(30000)).await;
+            Some((
+                match self::process_reporter_info(client.clone()).await {
+                    Ok((pods_strings, node_strings, cluster_stats_string)) => Some(
+                        stream::iter(
+                            pods_strings
+                                .into_iter()
+                                .chain(node_strings.into_iter())
+                                .chain(Some(cluster_stats_string).into_iter()),
+                        )
+                        .map(|line| {
+                            LineBuilder::new()
+                                .line(line)
+                                .file("logdna-reporter".to_string())
+                        }),
+                    ),
+                    Err(e) => {
+                        error!("Failed To Gather Metrics Server Info {}", e);
+                        None
```

Comment:
propagate the error up?

Reply:
this gets expanded on a bit in my next PR, around the leader election. 

My personal preference though as this stream just keeps running is to inform on the error, as its probably the metrics-server not running or a config issue, and keep trying every once in awhile. Rather than propagating up to the main agent flow, this feature is fairly niche and kind of an extra thing. 

The user also gets informed on the UI if something is wrong.